hubmap_api_py_client/external.py

```diff
@@ -76,36 +76,56 @@ def _operation(self, other_set, method):
             input_type=self.input_type, output_type=self.output_type,
             query=self.query)
 
-    def __getitem__(self, key):
-        if isinstance(key, int):
-            if key < 0:
-                raise ValueError('Negative index not supported')
-            return self.get_list(1, offset=key)[0]
-        raise TypeError('Use get_list for multiple values')
+    def get_list(self, values_included=[], sort_by=None):
+        return ResultsList(
+            client=self.client,
+            handle=self.handle,
+            input_type=self.input_type,
+            output_type=self.output_type,
+            values_included=values_included,
+            sort_by=sort_by)
 
-    def get_list(
-            self, limit, offset=0,
-            values_included=[], sort_by=None):
-        if not values_included and not sort_by:
-            return self.client.set_list_evaluation(
-                self.handle, self.output_type, limit,
-                offset=offset)
-        return self.client.set_detail_evaluation(
-            self.handle, self.output_type, limit,
-            offset=offset,
-            sort_by=sort_by,
-            values_type=self.input_type,
-            values_included=values_included)
 
-
-def class_name(output_type):
+def _class_name(output_type):
     return f'{output_type.capitalize()}ResultsSet'
 
 
 def _create_subclass(output_type):
-    return type(class_name(output_type), (ResultsSet,), {})
+    return type(_class_name(output_type), (ResultsSet,), {})
 
 
 for output_type in ['cell', 'organ', 'gene', 'cluster', 'dataset', 'protein']:
     ResultsSetSubclass = _create_subclass(output_type)
     _add_method(output_type, ResultsSetSubclass)
+
+
+class ResultsList():
+    def __init__(
+            self, client, handle,
+            input_type=None, output_type=None,
+            values_included=[], sort_by=None):
+        self.client = client
+        self.handle = handle
+        self.input_type = input_type
+        self.output_type = output_type
+        self.values_included = values_included
+        self.sort_by = sort_by
+
+    def __repr__(self):
+        return (
+            f'<ResultsList '
+            f'base_url={self.client.base_url} handle={self.handle} '
+            f'values_included={self.values_included} sort_by={self.sort_by}>')
+
+    def get(self, limit=10, offset=0):
+        # TODO: This will be replaced with a magic method.
+        if not self.values_included and not self.sort_by:
+            return self.client.set_list_evaluation(
+                self.handle, self.output_type, limit,
+                offset=offset)
+        return self.client.set_detail_evaluation(
+            self.handle, self.output_type, limit,
+            offset=offset,
+            sort_by=self.sort_by,
+            values_type=self.input_type,
```

Comment:
This line here is something I've been thinking about.  I appreciate the desire to not overwhelm the user with parameters, but I'm not sure this works.  For example, I think the API's design allows a user to do an organ-> cell query and then get gene data for the cells in that query, which wouldn't work here.  Additionally, as I understand it, when a ResultsSet is returned by an operation, the result has the input_type of the first operand, which I don't love.  I had considered trying to eliminate this parameter by coming up with a way to infer it based on the values_included/sort_by, and I think I just convinced myself that that's the best way to handle this, and that I should do so sooner rather than later.